Recent update on struts6-upg

kateyang1998 opened the pull request:
- Removed `PHR` module and fixed relevant codes.
- Cleaned up pom.xml (transitive / duplicated dependencies).
- Upgraded `axis2` from version 1.8.0 to 2.0.0-SNAPSHOT.
- Removed `myoscar_client_utils` library, fixed some compilation errors caused by removal.
- Replaced `org.apache.cxf.rs.security.oauth` to `scribejava`, refactored the relevant classes to make them compatible with `scribejava oauth1.0a`

yingbull reviewed (approved):
I am going to let this go forward but be sure that we have followed up to removed those other MyOSCAR items you added (and any others) and make sure an issue ticket tracks this till done.   Also, remember to not add commented out code unless there is a clear statement why this must be commented out and under what circumstances it would be uncommented/when.

kateyang1998 commented:
> This needs to be updated / merge issue. Please review @kateyang1998

Sure, I will look into it on Monday morning.

yingbull commented on src/main/java/com/onelogin/saml2/customize/authn/SamlResponse.java:
```diff
@@ -0,0 +1,1351 @@
+
+
+		// We need to Remove the saml:EncryptedAssertion Node
+		NodeList AssertionDataNodes = Util.query(dom, "/samlp:Response/saml:EncryptedAssertion/saml:Assertion");
```
@kateyang1998 ? Did you miss following up with sourcery on this?

yingbull commented:
@kateyang1998 This is also out of date, can you update what should be reviewed for merge?

kateyang1998 commented:
@yingbull I just updated the PR description, right now is up-to-date. The 20 commits are the latest changes we made to this branch. (struts6-upg)

yingbull commented on src/main/java/com/onelogin/saml2/customize/authn/SamlResponse.java:
```diff
@@ -0,0 +1,1351 @@
+	 * @throws ValidationError
+	 */
+	public void loadXmlFromBase64(String responseStr) throws ParserConfigurationException, XPathExpressionException, SAXException, IOException, SettingsException, ValidationError {
+		samlResponseString = new String(Util.base64decoder(responseStr), "UTF-8");
```
A good suggestion, but in this case we will leave it for minimal changes at this point, but this was correct to flag.

yingbull commented on src/main/java/oscar/login/OscarOAuthDataProvider.java:
```diff
@@ -24,200 +24,203 @@
-        long issuedAt = System.currentTimeMillis() / 1000;
-        AccessToken accessToken = new AccessToken(client, accessTokenString,
-                tokenSecretString, 3600, issuedAt);
-        UserSubject subject = new UserSubject(srt.getProviderNo(), new ArrayList<>());
```
Where are we capturing this level of detail now @kateyang1998 ?

yingbull commented on src/main/java/oscar/login/OscarOAuthDataProvider.java:
```diff
@@ -24,200 +24,203 @@
-        sat.setClientId(sc.getId());
-        sat.setDateCreated(new Date());
-        sat.setIssued(issuedAt);
-        sat.setLifetime(3600); // Assuming lifetime is 3600 seconds or should be set as per your application logic
```
What is token lifetime in this new setup?

yingbull commented on src/main/java/oscar/login/OscarOAuthDataProvider.java:
```diff
@@ -24,200 +24,203 @@
+    public OAuth1AccessToken getAccessToken(String tokenId) {
+        ServiceAccessToken sat = serviceAccessTokenDao.findByTokenId(tokenId);
         if (sat == null) {
-            throw new OAuthServiceException("Invalid access token.");
```
Similar comments to above.

yingbull commented on src/main/java/oscar/oscarEncounter/pageUtil/EctIncomingEncounter2Action.java:
```diff
@@ -192,13 +212,13 @@ public String execute() throws IOException, ServletException {
                 ResourceBundle props = ResourceBundle.getBundle("oscarResources", request.getLocale());
                 try {
-                    MessageTransfer3 messageTransfer = MyOscarMessagesHelper.readMessage(request.getSession(),
+                    MessageTransfer3 messageTransfer = readMessage(request.getSession(),
```
Why this change? If we aren't going myoscarmsgs, maybe this whole block goes away?

kateyang1998 replied:
This is because of myoscar dependency removal. After we removed the dependency, we had to do this refactor (The refactor of myoscar is not finished yet, we stopped and then switch to the 404/500 error fix in coyote)

yingbull commented on src/main/java/oscar/oscarEncounter/pageUtil/EctIncomingEncounter2Action.java:
```diff
@@ -192,13 +212,13 @@ public String execute() throws IOException, ServletException {
                     String dateStr = "";
 
                     if (request.getParameter("remyoscarmsg") != null) {
-                        MessageTransfer3 messageTransferOrig = MyOscarMessagesHelper.readMessage(request.getSession(),
```
As above.

kateyang1998 replied:
Same to above

yingbull commented on src/main/java/oscar/oscarLab/ca/all/pageUtil/InsideLabUpload2Action.java:
```diff
@@ -50,7 +50,6 @@
 
 import org.apache.commons.fileupload2.core.FileItem;
 import org.apache.commons.fileupload2.core.FileUploadException;
-import org.apache.commons.fileupload.servlet.ServletFileUpload;
```
What was this replaced with?

kateyang1998 replied:
This is the relpaced import:
import org.apache.commons.fileupload2.jakarta.servlet6.JakartaServletFileUpload;

yingbull commented on src/main/java/oscar/oscarRx/pageUtil/RxDeleteRx2Action.java:
```diff
@@ -219,7 +219,7 @@ public String clearPHRMeds()
             return null;
         }
-        bean.clearPairPHRMed();
+        //bean.clearPairPHRMed();
```
if not needed, let's remove, but feels like more code might be involved?

kateyang1998 commented:
@yingbull for the comments I haven't replied -
I will need to look into details, when I'm back, but these are some info from DeepSeek:
Q1: Are permissions being kept in how we are doing OAuth?
Yes, but handled differently:
Original:
Used OAuthPermission objects (Apache CXF)
Stored as space-separated strings in DB (e.g., "scope1 scope2")
Explicitly attached to tokens via setScopes()
Refactored:
Still stores scopes in ServiceRequestToken.scopes and ServiceAccessToken.scopes as space-separated strings.
But ScribeJava's OAuth1RequestToken/OAuth1AccessToken does not natively support scopes.
→ Recommendation:
If scopes are critical, add a Map<String, List<String>> in OAuth1Client to track token-scope relationships.
Or extend OAuth1AccessToken to include a scopes field.

Q2: Where are we capturing this level of detail now?
Scopes: Still stored in DB columns (ServiceRequestToken.scopes, ServiceAccessToken.scopes).
Missing in Refactored Code:
The refactored version does not attach scopes to OAuth1AccessToken when returned (unlike the original).
Fix: Modify createAccessToken() to include scopes:
java
public OAuth1AccessToken createAccessToken(...) {
    // ... existing code ...
    OAuth1AccessToken accessToken = new OAuth1AccessToken(token, secret);
    // Add scopes (requires custom subclass or external tracking)
    return accessToken;
}

Q3: What is token lifetime in this new setup?
Original:
Explicitly set lifetime=3600 (1 hour) for AccessToken.
Validated during token usage (e.g., getAccessToken() checks expiry).
Refactored:
No explicit lifetime in OAuth1AccessToken (ScribeJava does not enforce this).
Risk: Tokens never expire unless manually revoked.
Recommendation:
Add expiryDate to ServiceAccessToken and validate in getAccessToken():
java
public OAuth1AccessToken getAccessToken(String tokenId) {
    ServiceAccessToken sat = serviceAccessTokenDao.findByTokenId(tokenId);
    if (sat == null || sat.getExpiryDate().before(new Date())) {
        throw new RuntimeException("Invalid/expired token.");
    }
    return new OAuth1AccessToken(...);
}

Critical Missing Pieces in Refactored Code
Scope Handling:
Original: Attached scopes to tokens for granular permissions.
Refactored: Scopes are stored but not enforced.
Token Expiry:
Original: Hardcoded 1-hour lifetime.
Refactored: No expiry logic (security risk).
User Authorization:
Original: Used UserSubject to bind tokens to users (providerNo).
Refactored: Still stores providerNo but does not attach it to tokens.

Recommendations
Add Scope Support:
java
public class ScopedOAuth1AccessToken extends OAuth1AccessToken {
    private List<String> scopes;
    // ... getters/setters ...
}
Implement Token Expiry:
Add expiryDate to ServiceAccessToken.
Validate in getAccessToken().
Reattach User Context:
Extend OAuth1AccessToken to include providerNo or use a wrapper class.
Update Database Schema:
Add expiry_date column to service_access_token table.

kateyang1998 commented:
@yingbull I will start to look into this branch today, please let me know if there is any other issue you want me to work on first, or I will just focus on this PR now :-)

yingbull commented:
@sebastian-j-ibanez this PR might have some of the jakarta related changes that you might be able to point claude at to summarize as part of moving them to experimental/next update.
